sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/pubsub/PubsubIO.java

```diff
@@ -1390,11 +1413,25 @@ public void processElement(@Element PubsubMessage message, @Timestamp Instant ti
         PreparePubsubWriteDoFn.validatePubsubMessageSize(message, maxPublishBatchSize);
         byte[] payload = message.getPayload();
         int messageSize = payload.length;
+
+        PubsubTopic pubsubTopic;
+        if (getTopicProvider() != null) {
+          pubsubTopic = getTopicProvider().get();
+        } else {
+          pubsubTopic =
+              PubsubTopic.fromPath(Preconditions.checkArgumentNotNull(message.getTopic()));
+        }
         // Checking before adding the message stops us from violating max batch size or bytes
-        if (output.size() >= maxPublishBatchSize
-            || (!output.isEmpty()
-                && (currentOutputBytes + messageSize) >= maxPublishBatchByteSize)) {
-          publish();
+        List<OutgoingMessage> currentTopicOutput =
+            output.computeIfAbsent(pubsubTopic, t -> Lists.newArrayList());
+        long currentTopicOutputBytes = currentOutputBytes.getOrDefault(pubsubTopic, 0L);
+        if (currentTopicOutput.size() >= maxPublishBatchSize
+            || (!currentTopicOutput.isEmpty()
+                && (currentTopicOutputBytes + messageSize) >= maxPublishBatchByteSize)) {
+          publish(pubsubTopic, currentTopicOutput);
+          currentTopicOutput.clear();
+          currentTopicOutputBytes = 0;
+          currentOutputBytes.put(pubsubTopic, 0L);
```

Comment:
think the single map would make this simpler to since you would be updating the value in the entry directly

Reply:
done